### PR TITLE
Add the ability to deploy without truffle fixture

### PR DESCRIPTION
**Pull request by jeremyjams**
Out scope
- Deploy & setup ENS without hardhat-truffle

### CODE REVIEW

**Review by gfournieriExec — APPROVED** [review]
Thanks for this first iteration

File: deploy/0_deploy.ts [hunk 1]
```diff
@@ -41,82 +62,176 @@ const ensPublicResolver: PublicResolver = hre.artifacts.require(
...
+        ? deploymentOptions.token // token
+            ? deploymentOptions.token
+            : await new RLC__factory()
+                  .connect(owner)
+                  .deploy()
+                  .then((contract) => {
+                      contract.deployed();
+                      return contract.address;
+                  })
+        : ethers.constants.AddressZero; // native
```
Comment (zguesmi) [inline deploy/0_deploy.ts:86]:
```suggestion
    let rlcInstanceAddress = isTokenMode
        ? getOrDeployRlc(deploymentOptions, owner) // token
        : ethers.constants.AddressZero; // native
```
with
```ts
async function getOrDeployRlc(deploymentOptions, owner) {
    return deploymentOptions.token
        ??  await new RLC__factory()
            .connect(owner)
            .deploy()
            .then((contract) => {
                contract.deployed();
                return contract.address;
            });
}
```
Or something similar. I'm just trying to make it more readable.

File: deploy/0_deploy.ts [hunk 2]
```diff
@@ -41,82 +62,176 @@ const ensPublicResolver: PublicResolver = hre.artifacts.require(
...
+    return name.replace('__factory', '');
+}
+
+/**
+ * Deploy through a GenericFactory a contract [and optionally trigger call]
+ */
+async function deploy(contractFactory: ContractFactory, constructorArgs?: any[], call?: string) {
+    let bytecode = constructorArgs
+        ? contractFactory.getDeployTransaction(...constructorArgs).data
+        : contractFactory.getDeployTransaction().data;
```
Comment (zguesmi) [inline deploy/0_deploy.ts:209]:
```suggestion
    const args: any[] = constructorArgs ?? [];
    let bytecode = contractFactory.getDeployTransaction(...args).data
```
or shorter
```suggestion
    let bytecode = contractFactory.getDeployTransaction(...(constructorArgs ?? [])).data
```

File: deploy/0_deploy.ts [hunk 3]
```diff
@@ -41,82 +62,176 @@ const ensPublicResolver: PublicResolver = hre.artifacts.require(
...
+ */
+function getBaseNameFromContractFactory(contractFactory: any) {
+    const name = contractFactory.constructor.name;
+    return name.replace('__factory', '');
+}
+
+/**
+ * Deploy through a GenericFactory a contract [and optionally trigger call]
+ */
+async function deploy(contractFactory: ContractFactory, constructorArgs?: any[], call?: string) {
```
Comment (zguesmi) [inline deploy/0_deploy.ts:211]:
```suggestion
async function deployWithFactory(contractToDeploy: ContractFactory, constructorArgs?: any[], call?: string) {
```

File: deploy/0_deploy.ts [hunk 4]
```diff
@@ -41,82 +62,176 @@ const ensPublicResolver: PublicResolver = hre.artifacts.require(
...
+                  .createContractAndCall(bytecode, salt, call)
+                  .then((tx) => tx.wait())
+            : await genericFactoryInstance.createContract(bytecode, salt).then((tx) => tx.wait());
+    }
+    const contractName = getBaseNameFromContractFactory(contractFactory);
+    console.log(
+        `${contractName}: ${contractAddress} ${previouslyDeployed ? ' (previously deployed)' : ''}`,
+    );
+    await deployments.save(contractName, {
+        abi: [],
```
Comment (zguesmi) [inline deploy/0_deploy.ts:232]: Can we use `contractFactory.constructor.abi` here?
```suggestion
        abi: contractFactory.constructor.abi,
```

File: deploy/0_deploy.ts (hunk 1 above)
Comment (jeremyjams) [reply deploy/0_deploy.ts]: Sure, see [Extract getOrDeployRlc](https://github.com/iExecBlockchainComputing/PoCo/pull/58/commits/106414b08ef5e9cbdc9a22913a63c5d30c1f1fe8) !

File: deploy/0_deploy.ts (hunk 2 above)
Comment (jeremyjams) [reply deploy/0_deploy.ts]: Better thanks! See [Get deployment bytecode with or without constructor args in single line](https://github.com/iExecBlockchainComputing/PoCo/pull/58/commits/1ae0b12e1e4d913659a13572d9fc61de5d8915e9)

File: utils/FactoryDeployer.js [hunk 5]
```diff
@@ -141,4 +141,4 @@ class TruffleDeployer extends EthersDeployer {
     }
 }
 
-module.exports = { EthersDeployer, TruffleDeployer };
+module.exports = { EthersDeployer, TruffleDeployer, FACTORY };
```
[the remark this answers is not in the document]
Comment (jeremyjams) [reply utils/FactoryDeployer.js]: Ok, tell me if you like this one or not: [Only expose factory address](https://github.com/iExecBlockchainComputing/PoCo/pull/58/commits/969714911e27c824404906100d0a87fc54e04b91)?

File: deploy/0_deploy.ts [hunk 6]
```diff
@@ -41,82 +63,178 @@ const ensPublicResolver: PublicResolver = hre.artifacts.require(
...
+    for (let i = catCountBefore.toNumber(); i < categories.length; i++) {
+        const category = categories[i];
+        await IexecCategoryManager__factory.connect(erc1538ProxyAddress, owner).createCategory(
+            category.name,
+            JSON.stringify(category.description),
+            category.workClockTimeRef,
+        );
+    }
+    const catCountAfter = await iexecAccessorsInstance.countCategory();
+    console.log(`countCategory is now: ${catCountAfter}`);
```
Comment (zguesmi) [inline deploy/0_deploy.ts:181]:
```suggestion
    console.log(`countCategory is now: ${catCountAfter} (was ${catCountBefore})`);
```

File: deploy/0_deploy.ts (hunk 3 above)
Comment (jeremyjams) [reply deploy/0_deploy.ts]: Thanks, see [Update wording and add abi on save deployment](https://github.com/iExecBlockchainComputing/PoCo/pull/58/commits/c0df6b1d4f10f8dcf16f699947f12c40a46c3155)

File: deploy/0_deploy.ts (hunk 4 above)
Comment (jeremyjams) [reply deploy/0_deploy.ts]: Though the same, we don't use it, but let's add it :) [Update wording and add abi on save deployment](https://github.com/iExecBlockchainComputing/PoCo/pull/58/commits/c0df6b1d4f10f8dcf16f699947f12c40a46c3155)

File: utils/FactoryDeployer.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (zguesmi) [reply utils/FactoryDeployer.js]: Thans!

File: deploy/0_deploy.ts (hunk 4 above)
Comment (zguesmi) [reply deploy/0_deploy.ts]: If it's not used we can remove it. It's up to you.

File: deploy/0_deploy.ts [hunk 7]
```diff
@@ -41,82 +63,186 @@ const ensPublicResolver: PublicResolver = hre.artifacts.require(
...
+                  .createContractAndCall(bytecode, salt, call)
+                  .then((tx) => tx.wait())
+            : await genericFactoryInstance.createContract(bytecode, salt).then((tx) => tx.wait());
+    }
+    const contractName = getBaseNameFromContractFactory(contractFactory);
+    console.log(
+        `${contractName}: ${contractAddress} ${previouslyDeployed ? ' (previously deployed)' : ''}`,
+    );
+    await deployments.save(contractName, {
+        abi: (contractFactory as any).constructor.abi,
```
Comment (zguesmi) [inline deploy/0_deploy.ts:240]: Not very important but if it's not used we can also remove it. It's up to you.

**Review by zguesmi — APPROVED** [review]
Clean!

File: deploy/0_deploy.ts (hunk 7 above)
Comment (jeremyjams) [reply deploy/0_deploy.ts]: Added a comment to be sure: https://github.com/iExecBlockchainComputing/PoCo/pull/58/commits/60b6907887659cbb667a151b3846bf78489d3b70

File: utils/FactoryDeployerHelper.ts [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+    genericFactory: GenericFactory;
+
+    constructor(owner: SignerWithAddress, salt: string) {
+        this.salt = salt;
+        this.init = new Deployer(owner);
+        this.genericFactory = GenericFactory__factory.connect(factoryAddress, owner);
+    }
+
+    /**
+     * Deploy through a GenericFactory a contract [and optionally trigger call]
```
Comment (zguesmi) [inline utils/FactoryDeployerHelper.ts:23]:
```suggestion
     * Deploy a contract through GenericFactory [and optionally trigger a call]
```

File: utils/FactoryDeployerHelper.ts [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+            : await this.genericFactory.predictAddress(bytecode, this.salt);
+        const previouslyDeployed = (await ethers.provider.getCode(contractAddress)) !== '0x';
+        if (!previouslyDeployed) {
+            call
+                ? await this.genericFactory
+                      .createContractAndCall(bytecode, this.salt, call)
+                      .then((tx) => tx.wait())
+                : await this.genericFactory
+                      .createContract(bytecode, this.salt)
+                      .then((tx) => tx.wait());
```
Comment (zguesmi) [inline utils/FactoryDeployerHelper.ts:46]:
```suggestion
            const tx = call
                ? await this.genericFactory.createContractAndCall(bytecode, this.salt, call)
                : await this.genericFactory.createContract(bytecode, this.salt);
            tx.wait();
```

Comment (jeremyjams) [reply utils/FactoryDeployerHelper.ts]: Could idea to remove duplication, please check a slightly different (missing `await` in suggestion) [Factorize create & predict contract](https://github.com/iExecBlockchainComputing/PoCo/pull/58/commits/c01bc8f44b714562182cb0967b44e36f7bc123e8) :ear:

**Review by zguesmi — APPROVED** [review]

**Pull request merged**